secure/push-metrics/src/pusher.rs

```diff
@@ -0,0 +1,64 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::counters::Metrics;
+use libra_logger::*;
+use std::{env, sync::Arc, thread, thread::JoinHandle, time::Duration};
+
+const DEFAULT_PUSH_FREQUENCY_SECS: u64 = 15;
+
+/// MetricsPusher provides a function to push a list of Metrics to a configurable
+/// pushgateway endpoint.
+pub struct MetricsPusher {
+    metrics: Arc<dyn Metrics + Send + Sync>,
+}
+
+impl MetricsPusher {
+    pub fn new(metrics: Arc<dyn Metrics + Send + Sync>) -> Self {
+        Self { metrics }
+    }
+
+    /// start starts a new thread and periodically pushes the metrics to a pushgateway endpoint
+    pub fn start(self) -> JoinHandle<()> {
+        thread::spawn(move || {
+            // eg value for PUSH_METRICS_ENDPOINT: "http://pushgatewar.server.com:9091/metrics/job/safety_rules"
+            let push_metrics_endpoint = match env::var("PUSH_METRICS_ENDPOINT") {
+                Ok(s) => s,
+                Err(_) => {
+                    info!("PUSH_METRICS_ENDPOINT env var is not set. Skipping sending metrics.");
+                    return;
+                }
+            };
+            let push_metrics_frequency_secs = match env::var("PUSH_METRICS_FREQUENCY_SECS") {
+                Ok(s) => match s.parse::<u64>() {
+                    Ok(i) => i,
+                    Err(_) => {
+                        error!("Invalid value for PUSH_METRICS_FREQUENCY_SECS: {}", s);
+                        return;
+                    }
+                },
+                Err(_) => DEFAULT_PUSH_FREQUENCY_SECS,
+            };
+            info!(
+                "Starting push metrics loop. Sending metrics to {} with a frequency of {} seconds",
+                push_metrics_endpoint, push_metrics_frequency_secs
+            );
+            loop {
```

Comment:
let's move this into another, private, function called run

Reply:
Done